version: '3.9'
services:
  competence:
    container_name: Competence-APP
    build:
      context: ./app
      dockerfile: Dockerfile
    environment:
      - VITE_HOST_API=api
    networks:
      - nginx
    expose:
      - "8080"
    depends_on:
      api:
        condition: service_healthy

  database:
    container_name: Database
    image: mysql:latest
    restart: unless-stopped
    environment:
      - MYSQL_ROOT_PASSWORD=${DB_PASSWORD}
      - MYSQL_DATABASE=mydb
      - MYSQL_USER=${DB_USER}
      - MYSQL_PASSWORD=${DB_PASSWORD}
    volumes:
      - ./docker/dump:/docker-entrypoint-initdb.d
    ports:
      - "3306:3306"
    healthcheck:
      test: mysqladmin ping -h localhost -u ${DB_USER} -p${DB_PASSWORD}
      timeout: 10s
      interval: 3s
      start_period: 5s
      retries: 15
    expose:
      - "3306"
    networks:
      - database-services

  api:
    container_name: API
    build:
      context: ./api
      dockerfile: Dockerfile
    environment:
      - DB_HOST=database
      - DB_NAME=mydb
      - DB_USER=${DB_USER}
      - DB_PASSWORD=${DB_PASSWORD}
    depends_on:
      database:
        condition: service_healthy
    networks:
      - database-services
      - internal
      - nginx
    links:
      - database
    ports:
      - "3000:3000"

  nginx-http:
    container_name: Server-NGINX-HTTP
    image: nginx:alpine
    restart: unless-stopped
    configs:
      - source: nginx-http
        target: /etc/nginx/conf.d/default.conf
    ports:
      - "80:80"
    networks:
      - nginx
    depends_on:
      competence:
        condition: service_started
    links:
      - competence
    labels:
      - "group=reverse-proxy"
      - "security=high"

networks:
  nginx:
    driver: bridge
  database-services:
      driver: bridge
  internal:
    driver: bridge

configs:
  nginx-http:
    file: docker/dev/nginx/nginx-http.conf
